add general check for yaml on rcl_yaml_param_parser

iuhilnehc-ynos opened the pull request:
Fixes #301

Signed-off-by: Chen Lihui <Lihui.Chen@sony.com>

**Updated**: Besides fixing to not allow '//' in node name, adding the general check and some special rules about wildcards(including some new rules on https://github.com/ros2/design/pull/303)

fujitatomoya commented on rcl_yaml_param_parser/src/namespace.c:
```diff
@@ -16,6 +16,19 @@
+/// Validate a name whether it is a valid namespace
+///
+static rcutils_ret_t
+__validate_namespace(const char * name)
```
why it should not be dependent on `rmw`? could you elaborate a little bit? I think it would be better to use `rmw` to validate the namespace instead of having redundant code else where.

> I found that using rmw_validate_namespace has some limitations because there are some extra rules, such as "/**" (maybe more in the future), need to be parsed.

you mean https://github.com/ros2/rclcpp/issues/1265?

iuhilnehc-ynos replied:
Thanks for your questions, @fujitatomoya 

> why it should not be dependent on rmw?

"/**" is a special rule at `rclcpp`, it is invalid checked by `rmw_validate_namespace`.
If using `rmw_validate_namespace`, we need to add more extra checking for these special rules inside `__validate_namespace`, which seems not a requirement of this issue.

>> maybe more in the future

> you mean ros2/rclcpp#1265?

Yes.

jacobperron replied:
Though this solves the reported issue, I think we should do a more general check if the node namespace (and node name) are valid. I would do this using `rmw_validate_namespace` and `rmw_validate_node_name`, and handle the special cases for wildcards `*` and `**`.

I suggest implementing this function as follows:

1. Make a copy of the name to validate
2. Substitute substrings `*` and `**`  with arbitrary valid strings (e.g. `WILDCARD`)
3. Pass the modified name to `rmw_validate_namespace`

By using the rmw functions we are consistent with what it means to be a valid name.

What do you think?

iuhilnehc-ynos replied:
@jacobperron

> I suggest implementing this function as follows

Thanks for your suggestion. I'll update it later.

jacobperron reviewed (commented):
I think we should also validate the node name. If you don't want to add it to this PR, we do it in a follow-up.

iuhilnehc-ynos commented:
@jacobperron

After rebasing from `master`, I need to force push these commits to this branch.
Could you please help review it again?

jacobperron commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -404,6 +409,118 @@ rcutils_ret_t parse_value(
+__validate_name(const char * name, rcutils_allocator_t allocator)
+{
+  // special rules
+  if (strcmp(name, "/**") == 0) {
```
I think we could add additional rules for wildcards now (even if rclcpp and rclpy don't support them yet).
If you don't want to, I think it's worth adding a TODO here.

iuhilnehc-ynos replied:
> If you don't want to, I think it's worth adding a TODO here.

I have added a TODO there.

iuhilnehc-ynos replied:
@jacobperron

I found you had created a PR for wildcards design on https://github.com/ros2/design/pull/303.
so I'll add additional rules for wildcards later.

jacobperron reviewed (commented):
LGTM.

Could you also add a test for the `/**` case? Thanks!

iuhilnehc-ynos commented:
@jacobperron 

> Could you also add a test for the /** case? Thanks!

I have added the test and added additional rules for wildcards based on https://github.com/ros2/design/pull/303.
Could you review it again?

jacobperron commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -404,6 +409,128 @@ rcutils_ret_t parse_value(
+
+    char * namespace = NULL;
+    // Besides "*", node name also support /*.
+    if (0 == strcmp(node_name, "*") && '/' == name[separator_pos - name]) {
```
I think the colcon (`:`) following a namespace should implicitly act like a slash separate (`/`).
For example:

```yaml
/foo:
  /*:
    ros__parameters:
      ...
```

Should result in the fully-qualified name `/foo//*`, which is an invalid name.

Therefore, I think this case can be simplified to:

```c
char * namespace = rcutils_strndup(name, separator_pos - name, allocator);
```

As for the test, a valid parameters yaml file would be:

```yaml
/foo/*:
    ros__parameters:
      ...
```

iuhilnehc-ynos replied:
> Should result in the fully-qualified name /foo//*, which is an invalid name.

okay.

> As for the test, a valid parameters yaml file would be:
> ```
> /foo/*:
>     ros__parameters:
>       ...
> ```

Thank you.

iuhilnehc-ynos replied:
@jacobperron

I have a question, do we need to support the following setting?
```
/foo/**/bar:
  some_node:
     ros__parameters:
        ...
```

jacobperron replied:
I would say that example is valid following the matching rules on the design document. That being said, I don't know if it will be supported in the client libraries any time soon so we can probably revisit this case as a follow-up. WDYT?

jacobperron replied:
Nevermind, I see you added logic to support this. See my other comment https://github.com/ros2/rcl/pull/809#discussion_r504865010.

jacobperron commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -404,6 +409,147 @@ rcutils_ret_t parse_value(
+    }
+  }
+
+  if (0 != strcmp(node_name, "*") && 0 != strcmp(node_name, "**")) {
```
I don't think a node name set to `**` should be valid, since it implies it has zero or more parts (eg. `""` or `"/foo/bar"`).

iuhilnehc-ynos replied:
This case was for '/foo/bar/**'. OK, I'll add more codes to not allow node name to be '**'.

jacobperron replied:
I see. Yeah, this is tricky case. Maybe it's fine to leave the code as-is.

iuhilnehc-ynos replied:
okay.

jacobperron commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -404,6 +409,147 @@ rcutils_ret_t parse_value(
+        goto clean;
+      }
+    } else {
+      do {
```
I think we can refactor the following logic so it is easier to read and maintain. Instead of making repeated allocations in the loop, what if we just did a find-and-replace of all `*` characters in the string with a random valid character (e.g. `*` -> `W`)? Then pass the result to `__validate_namespace`. It seems like it would be a lot simpler to me. The only caveat is that we risk running over the limit `RMW_NAMESPACE_MAX_LENGTH`, but I'm not sure that is a case we should worry about handling.

iuhilnehc-ynos replied:
Actually, I thought that before. If it's partial matches for wildcards, such as '/foo/**part/node', this policy failed.

jacobperron replied:
Ah, yeah that makes sense. Nevermind then.

jacobperron reviewed (approved):
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12770)](http://ci.ros2.org/job/ci_linux/12770/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7729)](http://ci.ros2.org/job/ci_linux-aarch64/7729/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10483)](http://ci.ros2.org/job/ci_osx/10483/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12714)](http://ci.ros2.org/job/ci_windows/12714/)

iuhilnehc-ynos commented:
The `rcl_yaml_param_parser.cppcheck.error: syntaxError (src/parse.c:478)` on [macOS](http://ci.ros2.org/job/ci_osx/10483/) confused me, I'll rebuild it on my local environment tommorow because I have no environment now.

iuhilnehc-ynos commented:
@ros-pull-request-builder retest this please

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12779)](http://ci.ros2.org/job/ci_linux/12779/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7738)](http://ci.ros2.org/job/ci_linux-aarch64/7738/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10492)](http://ci.ros2.org/job/ci_osx/10492/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12725)](http://ci.ros2.org/job/ci_windows/12725/)

jacobperron commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -18,14 +18,54 @@
+/// \return RCUTILS_RET_OK if namespace is valid, or
+/// \return RCUTILS_RET_INVALID_ARGUMENT if namespace is not valid, or
+/// \return RCUTILS_RET_ERROR if an unspecified error occurred.
+RCL_YAML_PARAM_PARSER_LOCAL
```
We should include `rcl_yaml_param_parser/visibility_control.h`.

iuhilnehc-ynos replied:
OK. I'll add it.
(Actually, I thought rcl_yaml_param_parser/parser.h already includes rcl_yaml_param_parser/visibility_control.h, so I ignore it before.)

jacobperron replied:
Maybe it's redundant, but I think it's generally good to directly include whatever is being used in a file.

jacobperron commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -18,14 +18,54 @@
+/// \return RCUTILS_RET_ERROR if an unspecified error occurred.
+RCL_YAML_PARAM_PARSER_LOCAL
+rcutils_ret_t
+__validate_namespace(const char * namespace);
```
I think we should use a single underscore. Double underscores are reserved identifiers.
```suggestion
_validate_namespace(const char * namespace);
```

Same for other functions names below.

iuhilnehc-ynos replied:
That's good to know.

jacobperron reviewed (commented):
I'm not sure about the cppcheck failure on macOS or the Windows build failure :thinking:

iuhilnehc-ynos commented:
> cppcheck failure on macOS

the version of `cppcheck` on macOS is 1.89, so I should not use 'namespace' as a variable name which is a keyword in C++.

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12787)](http://ci.ros2.org/job/ci_linux/12787/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7746)](http://ci.ros2.org/job/ci_linux-aarch64/7746/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10501)](http://ci.ros2.org/job/ci_osx/10501/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12734)](http://ci.ros2.org/job/ci_windows/12734/)

jacobperron reviewed (approved):
LGTM, thanks for iterating!
